### ISSUE TITLE
[Incident] Diminished service for MEOM IGE / Grenoble hub

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/pilot-hubs#680]
# Summary

We've got a report of diminished service for the MEOM IGE hub. Since we use these incident reports for reporting what happened, I will try to keep use it for describing this incident and we can see how that goes.

Will add updates to the timeline below

### 2021-09-13 07:00AM US/Pacific

FreshDesk report received from @roxyboy, who reported instability / diminished service with the following description: "the notebooks don't load properly with "No Kernel" and/or get killed with the message: "Server unavailable or unreachable".

Going to meom-ige.2i2c.cloud allows me to log-in and authenticate as normal. Starting a server took a long time and I didn't see any log outputs. It looked like it was going to hang on this screen

![image](https://user-images.githubusercontent.com/1839645/133175931-488b4cea-ac6b-4dc3-8419-a26f521403db.png)

But then suddenly it popped up with JupyterLab as-expected. So I wasn't able to exactly reproduce the bug, but this might be something that intermittently happens.

Note that I can't find a grafana deployment at http://grafana.meom-ige.2i2c.cloud/ so it's hard for me to know what's going on without digging into the internals, so I haven't got much other information than this.

### 2021-09-13 00:00AM

@yuvipanda investigated and discovered that there was a race condition that likely caused the instability. See https://github.com/2i2c-org/pilot-hubs/issues/680#issuecomment-918954520 for debugging and assessment.

Opened up https://github.com/jupyterhub/kubespawner/pull/525 to fix this.

As the problem was intermittent, and a PR to fix the race condition has been filed, we'll consider this incident resolved.

---

(will fill in as we learn more)

### Technical improvements

1. This PR fixes the race condition that resulted in this problem: https://github.com/jupyterhub/kubespawner/pull/525

# Actions

- [x] Incident has been dealt with or is over
- [x] Sections above are filled out
- [x] All actionable items above have linked GitHub Issues

**Comment by choldgraf** [comment]
I think that this needs somebody from the @2i2c-org/tech-team to look at the hub or kubectl logs - it's hard for me to debug without Grafana :-/

**Comment by yuvipanda** [comment]
It seems to be working ok now - I see a few users on. 

For some reason, *node autoprovisioning* was on at some point in the past 2 days. It's off now. A node was brought up about 2 days ago (`node/gke-meom-ige-cluster-nap-e2-highmem-2-4cd17990-70pe`). 

In the hub logs, I see:

```
[W 2021-09-13 18:45:14.105 JupyterHub user:710] roxyboy's server failed to start in 600 seconds, giving up
[I 2021-09-13 18:45:14.106 JupyterHub spawner:2620] Deleting pod prod/jupyter-roxyboy
[W 2021-09-13 18:45:14.280 JupyterHub base:976] 3 consecutive spawns failed.  Hub will exit if failure count reaches 5 before succeeding
[E 2021-09-13 18:45:14.281 JupyterHub gen:623] Exception in Future <Task finished name='Task-192825' coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py:900> exception=TimeoutError('Timeout')> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 618, in error_callback
        future.result()
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 907, in finish_user_spawn
        await spawn_future
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/user.py", line 736, in spawn
        raise e
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/user.py", line 635, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
    tornado.util.TimeoutError: Timeout
    
[I 2021-09-13 18:45:14.282 JupyterHub log:189] 200 GET /hub/api/users/roxyboy/server/progress (roxyboy@10.128.0.58) 599719.01ms
[E 2021-09-13 18:46:09.873 JupyterHub spawner:2581] Pod prod/jupyter-roxyboy never showed up in reflector, restarting pod reflector
[E 2021-09-13 18:46:09.874 JupyterHub spawner:2585] Pods: {'prod/jupyter-auraoupa': {'metadata': {'name': 'jupyter-auraoupa', 'namespace': 'prod', 'selfLink': '/api/v1/namespaces/prod/pods/jupyter-auraoupa', 'uid': 'fa80b969-36eb-49a7-851e-6317150c4fbd', 'resourceVersion': '52693154', 'creationTimestamp': '2021-09-13T13:49:32Z', 'labels': {'app': 'jupyterhub', 'chart': 'jupyterhub-1.1.3', 'component': 'singleuser-server', 'heritage': 'jupyterhub', 'hub.jupyter.org/network-access-hub': 'true', 'hub.jupyter.org/network-access-proxy-http': 'true', 'hub.jupyter.org/servername': '', 'hub.jupyter.org/username': 'auraoupa', 'release': 'prod'}, 'annotations': {'hub.jupyter.org/username': 'auraoupa'}, 'managedFields': [{'manager': 'OpenAPI-Generator', 'operation': 'Update', 'apiVersion': 'v1', 'time': '2021-09-13T13:49:32Z', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:metadata': {'f:annotations': {'.': {}, 'f:hub.jupyter.org/username': {}}, 'f:labels': {'.': {}, 'f:app': {}, 'f:chart': {}, 'f:component': {}, 'f:heritage': {}, 'f:hub.jupyter.org/network-access-hub': {}, 'f:hub.jupyter.org/network-access-proxy-http': {}, 'f:hub.jupyter.org/servername': {}, 'f:hub.jupyter.org/username': {}, 'f:release': {}}}, 'f:spec': {'f:affinity': {'.': {}, 'f:nodeAffinity': {'.': {}, 'f:preferredDuringSchedulingIgnoredDuringExecution': {}}}, 'f:containers': {'k:{"name":"notebook"}': {'.': {}, 'f:args': {}, 'f:env': {'.': {}, 'k:{"name":"DASK_GATEWAY__ADDRESS"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"DASK_GATEWAY__AUTH__TYPE"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"DASK_GATEWAY__CLUSTER__OPTIONS__IMAGE"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"DASK_GATEWAY__PROXY_ADDRESS"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"DASK_GATEWAY__PUBLIC_ADDRESS"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"DATA_BUCKET"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JPY_API_TOKEN"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_ACTIVITY_URL"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_ADMIN_ACCESS"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_API_TOKEN"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_API_URL"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_BASE_URL"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_CLIENT_ID"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_HOST"}': {'.': {}, 'f:name': {}}, 'k:{"name":"JUPYTERHUB_OAUTH_CALLBACK_URL"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_SERVER_NAME"}': {'.': {}, 'f:name': {}}, 'k:{"name":"JUPYTERHUB_SERVICE_PREFIX"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTERHUB_USER"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTER_IMAGE"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTER_IMAGE_SPEC"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"JUPYTER_RUNTIME_DIR"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"MEM_GUARANTEE"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"MEM_LIMIT"}': {'.': {}, 'f:name': {}, 'f:value': {}}, 'k:{"name":"SCRATCH_BUCKET"}': {'.': {}, 'f:name': {}, 'f:value': {}}}, 'f:image': {}, 'f:imagePullPolicy': {}, 'f:lifecycle': {}, 'f:name': {}, 'f:ports': {'.': {}, 'k:{"containerPort":8888,"protocol":"TCP"}': {'.': {}, 'f:containerPort': {}, 'f:name': {}, 'f:protocol': {}}}, 'f:resources': {'.': {}, 'f:limits': {'.': {}, 'f:memory': {}}, 'f:requests': {'.': {}, 'f:memory': {}}}, 'f:securityContext': {'.': {}, 'f:allowPrivilegeEscalation': {}}, 'f:terminationMessagePath': {}, 'f:terminationMessagePolicy': {}, 'f:volumeMounts': {'.': {}, 'k:{"mountPath":"/home/jovyan"}': {'.': {}, 'f:mountPath': {}, 'f:name': {}, 'f:subPath': {}}, 'k:{"mountPath":"/home/jovyan/shared"}': {'.': {}, 'f:mountPath': {}, 'f:name': {}, 'f:readOnly': {}, 'f:subPath': {}}, 'k:{"mountPath":"/home/jovyan/shared-readwrite"}': {'.': {}, 'f:mountPath': {}, 'f:name': {}, 'f:subPath': {}}, 'k:{"mountPath":"/usr/local/etc/jupyter/jupyter_notebook_config.json"}': {'.': {}, 'f:mountPath': {}, 'f:name': {}, 'f:subPath': {}}}, 'f:workingDir': {}}}, 'f:dnsPolicy': {}, 'f:enableServiceLinks': {}, 'f:nodeSelector': {'.': {}, 'f:node.kubernetes.io/instance-type': {}}, 'f:priorityClassName': {}, 'f:restartPolicy': {}, 'f:schedulerName': {}, 'f:securityContext': {'.': {}, 'f:fsGroup': {}}, 'f:serviceAccount': {}, 'f:serviceAccountName': {}, 'f:terminationGracePeriodSeconds': {}, 'f:tolerations': {}, 'f:volumes': {'.': {}, 'k:{"name":"files"}': {'.': {}, 'f:name': {}, 'f:secret': {'.': {}, 'f:defaultMode': {}, 'f:items': {}, 'f:secretName': {}}}, 'k:{"name":"home"}': {'.': {}, 'f:name': {}, 'f:persistentVolumeClaim': {'.': {}, 'f:claimName': {}}}}}}}, {'manager': 'kube-scheduler', 'operation': 'Update', 'apiVersion': 'v1', 'time': '2021-09-13T13:49:32Z', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:status': {'f:conditions': {'.': {}, 'k:{"type":"PodScheduled"}': {'.': {}, 'f:lastProbeTime': {}, 'f:lastTransitionTime': {}, 'f:message': {}, 'f:reason': {}, 'f:status': {}, 'f:type': {}}}}}}, {'manager': 'kubelet', 'operation': 'Update', 'apiVersion': 'v1', 'time': '2021-09-13T13:52:09Z', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:status': {'f:conditions': {'k:{"type":"ContainersReady"}': {'.': {}, 'f:lastProbeTime': {}, 'f:lastTransitionTime': {}, 'f:status': {}, 'f:type': {}}, 'k:{"type":"Initialized"}': {'.': {}, 'f:lastProbeTime': {}, 'f:lastTransitionTime': {}, 'f:status': {}, 'f:type': {}}, 'k:{"type":"Ready"}': {'.': {}, 'f:lastProbeTime': {}, 'f:lastTransitionTime': {}, 'f:status': {}, 'f:type': {}}}, 'f:containerStatuses': {}, 'f:hostIP': {}, 'f:phase': {}, 'f:podIP': {}, 'f:podIPs': {'.': {}, 'k:{"ip":"10.0.2.2"}': {'.': {}, 'f:ip': {}}}, 'f:startTime': {}}}}]}, 'spec': {'volumes': [{'name': 'home', 'persistentVolumeClaim': {'claimName': 'home-nfs'}}, {'name': 'files', 'secret': {'secretName': 'singleuser', 'items': [{'key': 'jupyter_notebook_config.json', 'path': 'jupyter_notebook_config.json'}], 'defaultMode': 420}}, {'name': 'user-sa-token-rblf7', 'secret': {'secretName': 'user-sa-token-rblf7', 'defaultMode': 420}}], 'containers': [{'name': 'notebook', 'image': 'pangeo/pangeo-notebook:2021.07.17', 'args': ['jupyterhub-singleuser', '--ip=0.0.0.0', '--port=8888', '--SingleUserNotebookApp.default_url=/lab'], 'workingDir': '/home/jovyan', 'ports': [{'name': 'notebook-port', 'containerPort': 8888, 'protocol': 'TCP'}], 'env': [{'name': 'DASK_GATEWAY__ADDRESS', 'value': 'http://10.3.241.151:8000/services/dask-gateway/'}, {'name': 'DASK_GATEWAY__AUTH__TYPE', 'value': 'jupyterhub'}, {'name': 'DASK_GATEWAY__CLUSTER__OPTIONS__IMAGE', 'value': '{JUPYTER_IMAGE_SPEC}'}, {'name': 'DASK_GATEWAY__PROXY_ADDRESS', 'value': 'gateway://traefik-prod-dask-gateway.prod:80'}, {'name': 'DASK_GATEWAY__PUBLIC_ADDRESS', 'value': '/services/dask-gateway/'}, {'name': 'DATA_BUCKET', 'value': 'gcs://meom-ige-data'}, {'name': 'JPY_API_TOKEN', 'value': '88ccbb78f34f42d4b35d40a7ff34e30d'}, {'name': 'JUPYTERHUB_ACTIVITY_URL', 'value': 'http://hub:8081/hub/api/users/auraoupa/activity'}, {'name': 'JUPYTERHUB_ADMIN_ACCESS', 'value': '1'}, {'name': 'JUPYTERHUB_API_TOKEN', 'value': '88ccbb78f34f42d4b35d40a7ff34e30d'}, {'name': 'JUPYTERHUB_API_URL', 'value': 'http://hub:8081/hub/api'}, {'name': 'JUPYTERHUB_BASE_URL', 'value': '/'}, {'name': 'JUPYTERHUB_CLIENT_ID', 'value': 'jupyterhub-user-auraoupa'}, {'name': 'JUPYTERHUB_HOST'}, {'name': 'JUPYTERHUB_OAUTH_CALLBACK_URL', 'value': '/user/auraoupa/oauth_callback'}, {'name': 'JUPYTERHUB_SERVER_NAME'}, {'name': 'JUPYTERHUB_SERVICE_PREFIX', 'value': '/user/auraoupa/'}, {'name': 'JUPYTERHUB_USER', 'value': 'auraoupa'}, {'name': 'JUPYTER_IMAGE', 'value': 'pangeo/pangeo-notebook:2021.07.17'}, {'name': 'JUPYTER_IMAGE_SPEC', 'value': 'pangeo/pangeo-notebook:2021.07.17'}, {'name': 'JUPYTER_RUNTIME_DIR', 'value': '/tmp/.jupyter-runtime'}, {'name': 'MEM_GUARANTEE', 'value': '26843545600'}, {'name': 'MEM_LIMIT', 'value': '34359738368'}, {'name': 'SCRATCH_BUCKET', 'value': 'gcs://meom-ige-scratch/$(JUPYTERHUB_USER)'}], 'resources': {'limits': {'memory': '34359738368'}, 'requests': {'memory': '26843545600'}}, 'volumeMounts': [{'name': 'home', 'mountPath': '/home/jovyan', 'subPath': 'auraoupa'}, {'name': 'files', 'mountPath': '/usr/local/etc/jupyter/jupyter_notebook_config.json', 'subPath': 'jupyter_notebook_config.json'}, {'name': 'home', 'readOnly': True, 'mountPath': '/home/jovyan/shared', 'subPath': '_shared'}, {'name': 'home', 'mountPath': '/home/jovyan/shared-readwrite', 'subPath': '_shared'}, {'name': 'user-sa-token-rblf7', 'readOnly': True, 'mountPath': '/var/run/secrets/kubernetes.io/serviceaccount'}], 'lifecycle': {}, 'terminationMessagePath': '/dev/termination-log', 'terminationMessagePolicy': 'File', 'imagePullPolicy': 'IfNotPresent', 'securityContext': {'allowPrivilegeEscalation': False}}], 'restartPolicy': 'OnFailure', 'terminationGracePeriodSeconds': 30, 'dnsPolicy': 'ClusterFirst', 'nodeSelector': {'node.kubernetes.io/instance-type': 'e2-standard-8'}, 'serviceAccountName': 'user-sa', 'serviceAccount': 'user-sa', 'nodeName': 'gke-meom-ige-cluster-nb-medium-00e18158-q45d', 'securityContext': {'fsGroup': 100}, 'affinity': {'nodeAffinity': {'preferredDuringSchedulingIgnoredDuringExecution': [{'weight': 100, 'preference': {'matchExpressions': [{'key': 'hub.jupyter.org/node-purpose', 'operator': 'In', 'values': ['user']}]}}]}}, 'schedulerName': 'gke.io/optimize-utilization-scheduler', 'tolerations': [{'key': 'hub.jupyter.org/dedicated', 'operator': 'Equal', 'value': 'user', 'effect': 'NoSchedule'}, {'key': 'hub.jupyter.org_dedicated', 'operator': 'Equal', 'value': 'user', 'effect': 'NoSchedule'}, {'key': 'node.kubernetes.io/not-ready', 'operator': 'Exists', 'effect': 'NoExecute', 'tolerationSeconds': 300}, {'key': 'node.kubernetes.io/unreachable', 'operator': 'Exists', 'effect': 'NoExecute', 'tolerationSeconds': 300}], 'priorityClassName': 'prod-default-priority', 'priority': 0, 'enableServiceLinks': True, 'preemptionPolicy': 'PreemptLowerPriority'}, 'status': {'phase': 'Running', 'conditions': [{'type': 'Initialized', 'status': 'True', 'lastProbeTime': None, 'lastTransitionTime': '2021-09-13T13:50:41Z'}, {'type': 'Ready', 'status': 'True', 'lastProbeTime': None, 'lastTransitionTime': '2021-09-13T13:52:09Z'}, {'type': 'ContainersReady', 'status': 'True', 'lastProbeTime': None, 'lastTransitionTime': '2021-09-13T13:52:09Z'}, {'type': 'PodScheduled', 'status': 'True', 'lastProbeTime': None, 'lastTransitionTime': '2021-09-13T13:50:40Z'}], 'hostIP': '10.128.15.193', 'podIP': '10.0.2.2', 'podIPs': [{'ip': '10.0.2.2'}], 'startTime': '2021-09-13T13:50:41Z', 'containerStatuses': [{'name': 'notebook', 'state': {'running': {'startedAt': '2021-09-13T13:52:08Z'}}, 'lastState': {}, 'ready': True, 'restartCount': 0, 'image': 'docker.io/pangeo/pangeo-notebook:2021.07.17', 'imageID': 'docker.io/pangeo/pangeo-notebook@sha256:2736f8fca81bee4926a9bc2110679bcc171c86d0dde95458e86327f73ee8ebc5', 'containerID': 'containerd://1ecb202f4daf6e04f76eae0a87221b9e525e5d3849c720f012a3bcd2d21f94e5', 'started': True}], 'qosClass': 'Burstable'}}}
[I 2021-09-13 18:46:09.884 JupyterHub reflector:275] watching for pods with label selector='component=singleuser-server' in namespace prod
[E 2021-09-13 18:46:09.885 JupyterHub gen:623] Exception in Future <Task finished name='Task-192826' coro=<ConfiguratorSpawnerMixin.start() done, defined at /home/jovyan/.local/lib/python3.8/site-packages/jupyterhub_configurator/mixins.py:25> exception=TimeoutError('pod prod/jupyter-roxyboy did not start in 600 seconds!')> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 618, in error_callback
        future.result()
      File "/home/jovyan/.local/lib/python3.8/site-packages/jupyterhub_configurator/mixins.py", line 45, in start
        return await super().start(*args, **kwargs)
      File "/usr/local/lib/python3.8/dist-packages/kubespawner/spawner.py", line 2572, in _start
        await exponential_backoff(
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/utils.py", line 184, in exponential_backoff
        raise TimeoutError(fail_message)
    TimeoutError: pod prod/jupyter-roxyboy did not start in 600 seconds!
    
[I 2021-09-13 18:46:10.238 JupyterHub reflector:361] pods watcher stopped
[W 2021-09-13 18:46:10.238 JupyterHub reflector:363] pods watcher finished
```

I also see the reflector was restarted a few times:

```
[E 2021-09-13 18:30:19.281 JupyterHub spawner:2581] Pod prod/jupyter-roxyboy never showed up in reflector, restarting pod reflector
```

And I see connection errors to the k8s master:


```
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f8295270340>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f8294225b20>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f82942253d0>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
[E 2021-09-10 15:19:17.173 JupyterHub reflector:350] Error when watching resources, retrying in 0.2s
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/kubespawner/reflector.py", line 286, in _watch_and_update
        resource_version = self._list_and_update()
      File "/usr/local/lib/python3.8/dist-packages/kubespawner/reflector.py", line 227, in _list_and_update
        initial_resources = getattr(self.api, self.list_method_name)(**kwargs)
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api/core_v1_api.py", line 15302, in list_namespaced_pod
        return self.list_namespaced_pod_with_http_info(namespace, **kwargs)  # noqa: E501
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api/core_v1_api.py", line 15413, in list_namespaced_pod_with_http_info
        return self.api_client.call_api(
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 348, in call_api
        return self.__call_api(resource_path, method,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 180, in __call_api
        response_data = self.request(
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 373, in request
        return self.rest_client.GET(url,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/rest.py", line 239, in GET
        return self.request("GET", url,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/rest.py", line 212, in request
        r = self.pool_manager.request(method, url,
      File "/usr/local/lib/python3.8/dist-packages/urllib3/request.py", line 74, in request
        return self.request_encode_url(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/request.py", line 96, in request_encode_url
        return self.urlopen(method, url, **extra_kw)
      File "/usr/local/lib/python3.8/dist-packages/urllib3/poolmanager.py", line 375, in urlopen
        response = conn.urlopen(method, u.request_uri, **kw)
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 755, in urlopen
        retries = retries.increment(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/util/retry.py", line 574, in increment
        raise MaxRetryError(_pool, url, error or ResponseError(cause))
    urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='10.3.240.1', port=443): Max retries exceeded with url: /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f82942252b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
    
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f82940c6250>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f82942258b0>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f8294225730>: Failed to establish a new connection: [Errno 111] Connection refused')': /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server
[E 2021-09-10 15:19:17.891 JupyterHub reflector:350] Error when watching resources, retrying in 0.4s
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/kubespawner/reflector.py", line 286, in _watch_and_update
        resource_version = self._list_and_update()
      File "/usr/local/lib/python3.8/dist-packages/kubespawner/reflector.py", line 227, in _list_and_update
        initial_resources = getattr(self.api, self.list_method_name)(**kwargs)
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api/core_v1_api.py", line 15302, in list_namespaced_pod
        return self.list_namespaced_pod_with_http_info(namespace, **kwargs)  # noqa: E501
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api/core_v1_api.py", line 15413, in list_namespaced_pod_with_http_info
        return self.api_client.call_api(
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 348, in call_api
        return self.__call_api(resource_path, method,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 180, in __call_api
        response_data = self.request(
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/api_client.py", line 373, in request
        return self.rest_client.GET(url,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/rest.py", line 239, in GET
        return self.request("GET", url,
      File "/usr/local/lib/python3.8/dist-packages/kubernetes/client/rest.py", line 212, in request
        r = self.pool_manager.request(method, url,
      File "/usr/local/lib/python3.8/dist-packages/urllib3/request.py", line 74, in request
        return self.request_encode_url(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/request.py", line 96, in request_encode_url
        return self.urlopen(method, url, **extra_kw)
      File "/usr/local/lib/python3.8/dist-packages/urllib3/poolmanager.py", line 375, in urlopen
        response = conn.urlopen(method, u.request_uri, **kw)
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 783, in urlopen
        return self.urlopen(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 755, in urlopen
        retries = retries.increment(
      File "/usr/local/lib/python3.8/dist-packages/urllib3/util/retry.py", line 574, in increment
        raise MaxRetryError(_pool, url, error or ResponseError(cause))
    urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='10.3.240.1', port=443): Max retries exceeded with url: /api/v1/namespaces/prod/pods?fieldSelector=&labelSelector=component%3Dsingleuser-server (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f8294225610>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

So it looks like the k8s master had some trouble, leading to this fallout.

Looking at list of operations performed on the k8s cluster doesn't offer anything useful either:

```
gcloud container operations list                                                                                                                         2 ↵
NAME                              TYPE            LOCATION       TARGET                     STATUS_MESSAGE  STATUS  START_TIME                      END_TIME
operation-1630556818904-eeec976a  UPGRADE_NODES   us-central1-b  nap-e2-highmem-2-1fltnvlb                  DONE    2021-09-02T04:26:58.904363918Z  2021-09-02T04:30:31.970370918Z
operation-1630557058903-0229ffcf  UPGRADE_NODES   us-central1-b  nap-e2-highmem-2-1fltnvlb                  DONE    2021-09-02T04:30:58.903271928Z  2021-09-02T04:31:14.304552092Z
operation-1630567013193-8cd60786  REPAIR_CLUSTER  us-central1-b  meom-ige-cluster                           DONE    2021-09-02T07:16:53.193467877Z  2021-09-02T07:21:07.882408338Z
operation-1631287127600-e489f353  UPGRADE_MASTER  us-central1-b  meom-ige-cluster                           DONE    2021-09-10T15:18:47.60002144Z   2021-09-10T15:22:57.199257168Z
operation-1631384023631-5ff6d3d3  UPGRADE_NODES   us-central1-b  nap-e2-highmem-2-1fltnvlb                  DONE    2021-09-11T18:13:43.631296675Z  2021-09-11T18:17:24.16160192Z
```

**Comment by yuvipanda** [comment]
I also see:

```
[C 2021-09-10 15:20:08.913 JupyterHub spawner:2222] Pods reflector failed, halting Hub.
[C 2021-09-10 15:20:10.386 JupyterHub spawner:2222] Events reflector failed, halting Hub.
```

Except the hub did not actually halt! https://github.com/jupyterhub/kubespawner/blob/d49b75ab62f373fd0602cc8ce4290565fdec1582/kubespawner/spawner.py#L2223 is where the source is, and I see a `sys.exit(1)` there. If it had actually worked, the hub would've restarted, and self-recovered sooner (it did eventually self recover).

So this is a bug in kubespawner. A failed reflector also explains why there was no events being shown to you as your server spawned.

Apparently `sys.exit` in a thread (which is where the reflector is called from) does *not* actually exit the program - only the thread (https://stackoverflow.com/a/905224). This explains the issue.

I've restarted the hub pod manually to shake out any semi-broken reflectors. I see events are showing up normally now as I spawn a new server.

So short version is - we have encountered a bug in how kubespawner recovers from k8s master disruption.

**Comment by yuvipanda** [comment]
https://github.com/jupyterhub/kubespawner/pull/525 should fix this.

**Comment by roxyboy** [comment]
Thank you @yuvipanda !!

**Comment by choldgraf** [comment]
many thanks @yuvipanda for the triage here! I guess "restart the hub pod" is always a reasonable debugging step :-)

I'll close this, as I think it is resolved, I've updated the top comment, and we have a PR open to fix the underlying bug!
